File: TPII-2025-2/PadraoGoF/Estrutural/Bridge/Bridge01.js
// 1.1 - Implementação da Interface Cor
class Cor{
    constructor(cor){
        this.Cor = cor;
    }

    getCor(){
        return this.Cor;
    }
}

// 1.2 - Implementação Concreta de Cores:
class CorVermelha extends Cor {
    constructor() {
        super("Vermelho");
    }
}


class CorVerde extends Cor{
    constructor(){
        super("Verde");
    }
}

class CorAzul extends Cor{
    constructor(){
        super("Azul");
    }
}

// 2.1 - Implementação da Interface da Forma:
class Forma {
    constructor(cor) {
        this.cor = cor;
    }

    setCor(cor) {
        this.cor = cor;
    }

    desenhar() {
        throw new Error("Esse Metodo precisa ser implementado pela subclasse");
    }
}


// 2.2 - Abstração Refinada:
class Circulo extends Forma{
    desenhar(){
        console.log(`Desenhando um Circulo ${this.cor.getCor()}`);
    }
}

class Quadrado extends Forma{
    desenhar(){
        console.log(`Desenhando um Quadrado ${this.cor.getCor()}`);
    }
}

class Triangulo extends Forma{
    desenhar(){
        console.log(`Desenhando um Triangulo ${this.cor.getCor()}`);
    }
}

// 3 - Utilização - Cliente:
const vermelho = new CorVermelha();
const verde = new CorVerde();
const azul = new CorAzul();

const circulo1 = new Circulo(vermelho);
const circulo2 = new Circulo(verde);
const circulo3 = new Circulo(azul);

const quadrado1 = new Quadrado(vermelho);
const quadrado2 = new Quadrado(verde);
const quadrado3 = new Quadrado(azul);

const triangulo1 = new Triangulo(vermelho);
const triangulo2 = new Triangulo(verde);
const triangulo3 = new Triangulo(azul);

circulo1.desenhar();
circulo2.desenhar();
circulo3.desenhar();

quadrado1.desenhar();
quadrado2.desenhar();
quadrado3.desenhar();

triangulo1.desenhar();
triangulo2.desenhar();
triangulo3.desenhar();

File: TPII-2025-2/PadraoGoF/Estrutural/Bridge/Bridge01.py
# 1.1 - Implementação da Interface Cor
class Cor:
    def __init__(self, cor: str):
        self._cor = cor

    def get_cor(self):
        return self._cor


# 1.2 - Implementação Concreta de Cores:
class CorVermelha(Cor):
    def __init__(self):
        super().__init__("Vermelho")


class CorVerde(Cor):
    def __init__(self):
        super().__init__("Verde")


class CorAzul(Cor):
    def __init__(self):
        super().__init__("Azul")


# 2.1 - Interface/Abstração da Forma
class Forma:
    def __init__(self, cor: Cor):
        self.cor = cor

    def set_cor(self, cor: Cor):
        self.cor = cor

    def desenhar(self):
        raise NotImplementedError("O método desenhar() deve ser implementado pela subclasse")


# 2.2 - Abstrações Refinadas
class Circulo(Forma):
    def desenhar(self):
        print(f"Desenhando um Círculo {self.cor.get_cor()}")


class Quadrado(Forma):
    def desenhar(self):
        print(f"Desenhando um Quadrado {self.cor.get_cor()}")


class Triangulo(Forma):
    def desenhar(self):
        print(f"Desenhando um Triângulo {self.cor.get_cor()}")


# 3 - Utilização - Cliente
if __name__ == "__main__":
    vermelho = CorVermelha()
    verde = CorVerde()
    azul = CorAzul()

    circulo1 = Circulo(vermelho)
    circulo2 = Circulo(verde)
    circulo3 = Circulo(azul)

    quadrado1 = Quadrado(vermelho)
    quadrado2 = Quadrado(verde)
    quadrado3 = Quadrado(azul)

    triangulo1 = Triangulo(vermelho)
    triangulo2 = Triangulo(verde)
    triangulo3 = Triangulo(azul)

    circulo1.desenhar()
    circulo2.desenhar()
    circulo3.desenhar()

    quadrado1.desenhar()
    quadrado2.desenhar()
    quadrado3.desenhar()

    triangulo1.desenhar()
    triangulo2.desenhar()
    triangulo3.desenhar()
